File: Problem21.java
// Time Complexity: O(n)
// Space Complexity: O(n)
// House Robber

class Solution {
    public int rob(int[] nums) {
        // Edge case: if there are no houses, the maximum money robbed is 0
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int n = nums.length;

        // Edge case: if there is only one house, return the money in that house
        if (n == 1) {
            return nums[0];
        }

        // Initialize a dp array to store the maximum money that can be robbed up to
        // each house
        int[] dp = new int[n];

        // Base cases:
        // The maximum money that can be robbed from the first house is the money in the
        // first house
        dp[0] = nums[0];
        // The maximum money that can be robbed from the first two houses is the maximum
        // of the two houses
        dp[1] = Math.max(nums[0], nums[1]);

        // Fill the dp array using the recurrence relation
        for (int i = 2; i < n; i++) {
            // Either do not rob the current house and take the maximum up to the previous
            // house,
            // or rob the current house and add its money to the maximum up to two houses
            // back
            dp[i] = Math.max(dp[i - 1], nums[i] + dp[i - 2]);
        }

        // The answer is the maximum money that can be robbed up to the last house
        return dp[n - 1];
    }
}

File: problem 21.py
# Problem21: (https://leetcode.com/problems/house-robber/)

# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO


# Your code here along with comments explaining your approach


class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums) 
        if n == 0:
            return 0
        dp = [0 for _ in range(n+1)] # DP list of length n+1
        dp[1] = nums[0] # assign dp[1] as nums[0]
        for i in range(1,n): # for a loop from 1 to n
            val = nums[i]
            dp[i+1] = max(dp[i], (dp[i-1] + val)) # for the next entry either take the current value or 
            # take the previous entry + curr value
        
        return dp[n]
        